cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/OperationsDashboardHttpHandler.java

```diff
@@ -77,36 +109,104 @@ public void readDashboardDetail(FullHttpRequest request, HttpResponder responder
+    // if the end time is in the future, also add scheduled program runs to the result
+    if (endTimeSecs > TimeUnit.MICROSECONDS.toSeconds(System.currentTimeMillis())) {
+      // get enabled time schedules from all given namespaces
+      getEnabledTimeSchedules(namespaceIds)
+        // for each schedule, add all the scheduled runs within given the time range to the result
+        .forEach(schedule -> {
+          try {
+            result.addAll(getScheduledDashboardRecords(schedule, startTimeSecs, endTimeSecs));
+          } catch (Exception e) {
+            LOG.error("Failed to get scheduled program runs for schedule {}", schedule, e);
+          }
+        });
+    }
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(result));
+  }
+
+  /**
+   * Gets all the enabled time schedules in the given namespaces
+   */
+  private Stream<ProgramScheduleRecord> getEnabledTimeSchedules(Set<NamespaceId> namespaceIds) {
+    return namespaceIds.stream()
+      // get schedules from each namespace
+      .flatMap(ns -> scheduler.listScheduleRecords(ns).stream())
+      // filter schedules to get enabled time schedules
+      .filter(scheduleRecord ->
+      ProgramScheduleStatus.SCHEDULED.equals(scheduleRecord.getMeta().getStatus()) &&
+        Trigger.Type.TIME.equals(scheduleRecord.getSchedule().getTrigger().getType()));
   }
 
+  /**
+   * For a given time schedule, gets all the scheduled program run in the given time range
+   *
+   * @param scheduleRecord the schedule to get scheduled program run with
+   * @param startTimeSecs the start of the time range in seconds (inclusive, i.e. scheduled time larger or
+   *                      equal to the start will be returned)
+   * @param endTimeSecs the end of the time range in seconds (exclusive, i.e. scheduled time smaller than the end
+   *                    will be returned)
+   * @return a list of dashboard program run records with scheduled time as start time
+   * @throws Exception
+   */
+  private List<DashboardProgramRunRecord> getScheduledDashboardRecords(ProgramScheduleRecord scheduleRecord,
+                                                                       long startTimeSecs, long endTimeSecs)
+    throws Exception {
+    ProgramSchedule schedule = scheduleRecord.getSchedule();
+    ProgramId programId = schedule.getProgramId();
+    ProgramDescriptor programDescriptor = store.loadProgram(programId);
+
+    ArtifactSummary artifactSummary = ArtifactSummary.from(programDescriptor.getArtifactId().toApiArtifactId());
+    // if the program has a namespace user configured then set that user in the security request context.
+    // See: CDAP-7396
+    String nsPrincipal = namespaceQueryAdmin.get(programId.getNamespaceId()).getConfig().getPrincipal();
+    String userId = nsPrincipal != null && SecurityUtil.isKerberosEnabled(cConf) ?
+      new KerberosName(nsPrincipal).getServiceName() : null;
```

Comment:
not sure if this is the right approach or if needed, this looks expensive.

Reply:
the properties don't contain the principal. I can add it, but should we just return null for the older schedules?